### ISSUE TITLE
t/040_special_vars.t fails after reworked stringification of POK/NOK/IOK values

### ISSUE THREAD

**Issue by ikruglov** [Sereal/Sereal#111]
```
$ perl -I blib/lib/ -I blib/arch/ t/040_special_vars.t
1..6
ok 1 - ref undef
ok 2 - ref undef var
ok 3 - ref false
not ok 4 - ref false var
#   Failed test 'ref false var'
#   at t/040_special_vars.t line 40.
#          got: 'false'
#     expected: 'not-special'
ok 5 - ref true
ok 6 - ref true var
# Looks like you failed 1 test of 6.
```

Failing data structure is:
```
   [ "ref false var", \do { my $z = !1 },                                  ],
```

I quickly looked at the issue. The symptom is that Sereal::Encoder used to encode '\do { my $z = !1 }' as ref to empty string:

```
$ cat test | perl author_tools/hobodecoder.pl
Loaded constants from /Users/ikruglov/src/perl/Sereal/Perl/Decoder/blib/lib/Sereal/Decoder/Constants.pm
0000000    61 243 114 108   3   0  40  96
           = 363   r   l 003  \0   (   `
0000010


Total length: 8

Sereal protocol version: 3
Empty Header.
0 50 3
Header says: Document body is uncompressed.
000006/000001: 28  040 REFN
000007/000002: 60  096   SHORT_BINARY(0): ''
```

After e8c1309 encoder produce REFN FALSE tags:
```
$ cat test | perl author_tools/hobodecoder.pl
Loaded constants from /Users/ikruglov/src/perl/Sereal/Perl/Decoder/blib/lib/Sereal/Decoder/Constants.pm
0000000    61 243 114 108   3   0  40  58
           = 363   r   l 003  \0   (   :
0000010


Total length: 8

Sereal protocol version: 3
Empty Header.
0 50 3
Header says: Document body is uncompressed.
000006/000001: 28  040 REFN
000007/000002: 3a  058   FALSE

```

**Comment by demerphq** [comment]
On 8 November 2015 at 15:24, Ivan Kruglov <notifications@github.com> wrote:

> $ perl -I blib/lib/ -I blib/arch/ t/040_special_vars.t
> 1..6
> ok 1 - ref undef
> ok 2 - ref undef var
> ok 3 - ref false
> not ok 4 - ref false var
> #   Failed test 'ref false var'
> #   at t/040_special_vars.t line 40.
> #          got: 'false'
> #     expected: 'not-special'
> ok 5 - ref true
> ok 6 - ref true var
> # Looks like you failed 1 test of 6.
>
> Failing data structure is:
>
>    [ "ref false var", \do { my $z = !1 },                                  ],
>
> I quickly looked at the issue. The symptom is that Sereal::Encoder used to
> encode '\do { my $z = !1 }' as ref to empty string:
>
> $ cat test | perl author_tools/hobodecoder.pl
> Loaded constants from /Users/ikruglov/src/perl/Sereal/Perl/Decoder/blib/lib/Sereal/Decoder/Constants.pm
> 0000000    61 243 114 108   3   0  40  96
>            = 363   r   l 003  \0   (   `
> 0000010
>
>
> Total length: 8
>
> Sereal protocol version: 3
> Empty Header.
> 0 50 3
> Header says: Document body is uncompressed.
> 000006/000001: 28  040 REFN
> 000007/000002: 60  096   SHORT_BINARY(0): ''
>
> After e8c1309
> <https://github.com/Sereal/Sereal/commit/e8c1309c697766fe600575e96d4cdb7644357ba3>
> encoder produce REFN FALSE tags:
>
> $ cat test | perl author_tools/hobodecoder.pl
> Loaded constants from /Users/ikruglov/src/perl/Sereal/Perl/Decoder/blib/lib/Sereal/Decoder/Constants.pm
> 0000000    61 243 114 108   3   0  40  58
>            = 363   r   l 003  \0   (   :
> 0000010
>
>
> Total length: 8
>
> Sereal protocol version: 3
> Empty Header.
> 0 50 3
> Header says: Document body is uncompressed.
> 000006/000001: 28  040 REFN
> 000007/000002: 3a  058   FALSE
>
>
This is my bad. I know what the problem is, and I have more or less sorted
it out, but while I was testing I found a second related issues that I need
to sort out first.

Yves

**Comment by demerphq** [comment]
This is fixed in the 3.005_005/3.006 release. Thanks for the report.
